pkg/storage/ingest/reader.go

```diff
@@ -437,22 +408,73 @@ func (r *PartitionReader) newKafkaReader(at kgo.Offset) (*kgo.Client, error) {
 	return client, nil
 }
 
-func (r *PartitionReader) fetchLastCommittedOffsetWithRetries(ctx context.Context) (offset int64, exists bool, err error) {
-	var (
-		retry = backoff.New(ctx, backoff.Config{
-			MinBackoff: 100 * time.Millisecond,
-			MaxBackoff: 2 * time.Second,
-			MaxRetries: 10,
-		})
-	)
+func (r *PartitionReader) getStartOffset(ctx context.Context) (startOffset, lastConsumedOffset int64, err error) {
+	switch r.kafkaCfg.ConsumeFromPositionAtStartup {
+	case consumeFromStart:
+		startOffset = kafkaOffsetStart
+		level.Info(r.logger).Log("msg", "starting consumption from partition start", "start_offset", startOffset, "consumer_group", r.consumerGroup)
+		return startOffset, -1, nil
+
+	case consumeFromEnd:
+		startOffset = kafkaOffsetEnd
+		level.Warn(r.logger).Log("msg", "starting consumption from partition end (may cause data loss)", "start_offset", startOffset, "consumer_group", r.consumerGroup)
+		return startOffset, -1, nil
+	}
+
+	// We use an ephemeral client to fetch the offset and then create a new client with this offset.
+	// The reason for this is that changing the offset of an existing client requires to have used this client for fetching at least once.
+	// We don't want to do noop fetches just to warm up the client, so we create a new client instead.
+	cl, err := kgo.NewClient(commonKafkaClientOptions(r.kafkaCfg, r.metrics.kprom, r.logger)...)
+	if err != nil {
+		return 0, -1, fmt.Errorf("unable to create bootstrap client: %w", err)
+	}
+	defer cl.Close()
 
+	// lastConsumedOffset is unknown until (and only if) we successfully fetch its value below.
+	lastConsumedOffset = -1
+
+	fetchOffset := func(ctx context.Context) (offset int64, err error) {
+		if r.kafkaCfg.ConsumeFromPositionAtStartup == consumeFromTimestamp {
+			ts := time.UnixMilli(r.kafkaCfg.ConsumeFromTimestampAtStartup)
+			offset, exists, err := r.fetchFirstOffsetAfterTime(ctx, cl, ts)
+			if err != nil {
+				return 0, err
+			}
+			if exists {
+				level.Info(r.logger).Log("msg", "starting consumption from timestamp", "timestamp", ts.UnixMilli(), "start_offset", offset, "consumer_group", r.consumerGroup)
```

Comment:
I have a doubt about an edge case here. The `lastConsumedOffset` is used `processNextFetchesUntilMaxLagHonored()` to check whether the last produced offset has been consumed, and so it reader catched up. If the configured timestamp position is the end of the partition, and none is writing to the partition anymore, `processNextFetchesUntilMaxLagHonored()` may never return because `lastConsumedOffset` has not been initialised (we return -1 here) and, at the same time, it will be never set because no new records are produced. I'm wondering if, instead, we should simply initialise `lastConsumedOffset` to `offset - 1` (that is the offset before the one we'll start the consumption from).

For the other cases (consume from start/end/last-offset) this was an handled case (e.g. see the "should not wait indefinitely if there are no records to consume from Kafka but partition start offset is > 0 (e.g. all previous records have been deleted by Kafka retention)" test case, which is an edge case we've hit in dev when restarting an ingester consuming from a scaled-down partition).

Reply:
Good point, I've pushed a commit, that addresses that. PHAL